### PR TITLE
Add spec for ReverseOp

### LINKED ISSUE #247: Add spec for ReverseOp

### PR DESCRIPTION
**Pull request by subhankarshah**
Closes #247

### CODE REVIEW

File: docs/spec_draft.md [hunk 1]
```diff
@@ -927,6 +928,47 @@ spaces of `result` and `operand`.
...
+
+### Results
+
+| Name     | Type                          |
+|----------|-------------------------------|
+| `result` | tensor of any supported types |
+
+### Constraints
+
+  * (C1) `operand` and `result` have the same element type.
```
Comment (sdasgup3) [inline docs/spec_draft.md:952]: Combine C1 and C2 to say that they have the same type.

File: docs/spec_draft.md [hunk 2]
```diff
@@ -927,6 +928,47 @@ spaces of `result` and `operand`.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.reverse
+
+### Semantics
+
+Reverses the order of elements in the `operand` along the specified `dimensions`, generating an output array of the same shape. Each element of the operand array at a multidimensional index is stored into the output array at a transformed index. The multidimensional index is transformed by reversing the index in each dimension to be reversed (i.e., if a dimension of size N is one of the reversing dimensions, its index i is transformed into N - 1 - i).
```
Comment (sdasgup3) [inline docs/spec_draft.md:935]: Similar to how we are writing the transpose 


result[i0, ..., ik,..., iR-1] = operand[i0, ..., ik',..., iR-1] such that ik + ik' = dk - 1

File: docs/spec_draft.md [hunk 3]
```diff
@@ -167,6 +167,7 @@ described below)
    * [or](#stablehloor)
```
Comment (burmako) [inline docs/spec_draft.md:167]: Also align prose to 80 characters (see other sections in the file). Examples can be longer than 80 characters. We don't have a linter for this yet, but this is what we're informally following in the spec draft.

File: docs/spec_draft.md (hunk 2 above)
Comment (burmako) [inline docs/spec_draft.md:935]: "generating an output array of the same shape". So far, the convention has been to write this as "and produces a \`result\` tensor".

Comment (burmako) [reply docs/spec_draft.md:935]: +1, with a small change: `result[i0, ..., ik,..., iR-1] = operand[i0, ..., ik',..., iR-1]` where `ik + ik' = dk - 1` **for all dimensions `k` in `dimensions`**.

File: docs/spec_draft.md (hunk 1 above)
Comment (burmako) [inline docs/spec_draft.md:952]: Additionally:

```
* (C3) All dimensions in `dimensions` are unique.
* (C4) For all dimensions in `dimensions`, 0 $\le$ `dimension` $\lt$ `R`, where `R` is the rank of the `result`.
```

The checklist mentioned above provides links to files which contain existing implementations of constraints for HLO and StableHLO. It's a good idea to derive constraints from there.

File: docs/spec_draft.md [hunk 4]
```diff
@@ -927,6 +928,47 @@ spaces of `result` and `operand`.
...
+### Constraints
+
+  * (C1) `operand` and `result` have the same element type.
+  * (C2) `operand` and `result` have the same shape.
+
+### Examples
+
+```mlir
+// %operand = [[1, 2], [3, 4], [5, 6]]
+%result = "stablehlo.reverse"(%operand) {dimensions = dense<[0]> : tensor<1xi64>} : (tensor<3x2xi32>) -> tensor<3x2xi32>
```
Comment (burmako) [inline docs/spec_draft.md:959]: Can be simplified to `dense<0>`. Same below.

**Review by burmako — CHANGES_REQUESTED** [review]

File: docs/spec_draft.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (burmako) [reply docs/spec_draft.md:167]: I propose that we update these files now, so that reviewers look at this pull request in its fullness the entire time. 

Sure we might need a rebase before merging, but that'll be a one-time thing, right? It's not like rebases are needed every time there's an update to a pull request (the updates can be simply pushed into the PR branch). 

Overall, I feel that the benefit that we'll be getting from more thorough reviews exceeds the additional effort needed to maintain status.md and StablehloOps.td in PRs like this one.

[the remark this answers is not in the document]
Comment (subhankarshah) [reply docs/spec_draft.md:167]: Sure

File: docs/spec_draft.md (hunk 4 above)
Comment (ghpvnist) [reply docs/spec_draft.md:959]: On a related note, can be simplified to `tensor<i64>` for scalars.

Nit: add space between the attribute and the curly braces?
```
%result = "stablehlo.reverse"(%operand) { dimensions = dense<0> : tensor<i64> } : (tensor<3x2xi32>) -> tensor<3x2xi32>
```

**Review by ghpvnist — CHANGES_REQUESTED** [review]

File: docs/spec_draft.md (hunk 4 above)
Comment (burmako) [reply docs/spec_draft.md:959]: Good idea about `tensor<i64>`! Didn't know that would work too.

Re: spacing, that would work too, but that's not the canonical form of the MLIR generic syntax (the canonical form doesn't have spacing).

Comment (ghpvnist) [reply docs/spec_draft.md]: Re: spacing, does the canonical form allow multi-line? This would be hard to follow on examples with bunch of attributes like `gather`.

File: stablehlo/dialect/StablehloOps.td [hunk 5]
```diff
@@ -2360,10 +2360,7 @@ def StableHLO_ReverseOp: StableHLO_Op<"reverse",
       [NoSideEffect, HLO_CompatibleOperandsAndResultType]> {
   let summary = "Reverse operator";
   let description = [{
-    Reverses the specified dimensions of `operand` according to the given
-    `dimensions`.
-
-    See https://www.tensorflow.org/xla/operation_semantics#rev_reverse.
+    https://github.com/openxla/stablehlo/blob/main/docs/spec_draft.md#stablehloreverse
```
Comment (ghpvnist) [inline stablehlo/dialect/StablehloOps.td:2363]: To be consistent with nearby PR, change to:

```
Reverses the order of elements in the `operand` along the specified 
`dimensions` and produces a `result` tensor.

See:
https://github.com/openxla/stablehlo/blob/main/docs/spec_draft.md#stablehloreverse

Example:
%0 = "stablehlo.reverse"(%arg0) {dimensions = dense<1> : tensor<i64>} : (tensor<3x2xi32>) -> tensor<3x2xi32>
```

File: docs/status.md [hunk 6]
```diff
@@ -112,7 +112,7 @@ one of the following tracking labels.
 | replica_id               |      no       |     yes*     | yes(need-revisit) |        yes        |     no      |
 | reshape                  |      yes      |     yes      |    infeasible     |        yes        |     yes     |
 | return                   |      no       |     yes*     |        no         |        yes        |     no      |
-| reverse                  |      no       |     yes*     |       yes*        |        no         |     no      |
+| reverse                  |      yes      |     yes*     |       yes*        |        no         |     no      |
```
Comment (ghpvnist) [inline docs/status.md:115]: Have you checked the verification and type inference columns? If they are not in sync, let's create a ticket following the [protocol](https://github.com/openxla/stablehlo/blob/main/docs/spec_checklist.md).

File: docs/spec_draft.md [hunk 7]
```diff
@@ -931,6 +932,55 @@ spaces of `result` and `operand`.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.reverse
+
+### Semantics
+
+Reverses the order of elements in the `operand` along the specified
```
Comment (ghpvnist) [inline docs/spec_draft.md:939]: Overall lgtm. Some changes to match the wording of the rest of the spec. Wdyt?

```
Reverses the order of elements in the `operand` along the specified `dimensions`
and produces a `result` tensor. Each element of `operand` tensor at each index
is stored into the `result` tensor such that the indices at `dimensions[k]` of
`operand` elements are reversed for all dimensions `k` in `dimensions`. More
formally, `result[i0, ..., ik,..., iR-1] = operand[i0, ..., ik',..., iR-1]`
where `ik + ik' = dk - 1` for all dimensions `k` in `dimensions`.
```

**Review by ghpvnist — CHANGES_REQUESTED** [review]

File: docs/status.md (hunk 6 above)
Comment (subhankarshah) [reply docs/status.md:115]: The verification and type inference are in sync with the spec draft, I should make them yes according to the protocol but I did not see any other values in these columns with a yes, which makes me wonder if I am missing anything.

File: stablehlo/dialect/StablehloOps.td (hunk 5 above)
Comment (subhankarshah) [reply stablehlo/dialect/StablehloOps.td:2363]: Sounds good.

File: docs/spec_draft.md [hunk 8]
```diff
@@ -931,6 +932,55 @@ spaces of `result` and `operand`.
...
+### Constraints
+
+  * (C1) `operand` and `result` have the same type.
+  * (C2) All dimensions in `dimensions` are unique.
+  * (C3) For all dimensions in `dimensions`, 0 $\le$ `dimension` $\lt$ `R`,
+         where `R` is the rank of the `result`.
+
+### Examples
+
+```mlir
```
[the remark this answers is not in the document]
Comment (subhankarshah) [reply docs/spec_draft.md:969]: Looks good, changed 
`1-dimensional reverse` to `Reverse along dimension 0`
`2-dimensional reverse` to `Reverse along dimension 1`

File: docs/spec_draft.md (hunk 7 above)
Comment (subhankarshah) [reply docs/spec_draft.md:939]: Reverses the order of elements in the `operand` along the specified `dimensions`
and produces a `result` tensor.  For every dimension `k` in `dimensions`, an element at index `[i0, ..., ik,..., iR-1]` is swapped with the element at index `[i0, ..., dk-ik-1,..., iR-1]`, for all elements of the operand, thus reversing along the dimension `k`. 
More formally, after reversing along all dimensions, `result[i0, ..., ik,..., iR-1] = operand[i0, ..., ik',..., iR-1]`
where `ik + ik' = dk - 1` for all dimensions `k` in `dimensions`.

Comment (subhankarshah) [reply docs/spec_draft.md:939]: How about this.

Comment (burmako) [reply docs/spec_draft.md:939]: Can we drop the second sentence? I don't think it adds much for the first and third sentences, but the mental overhead is significant.

"Reverses the order of elements in the `operand` along the specified `dimensions` and produces a result tensor. More formally, `result[i0, ..., ik,..., iR-1] = operand[i0, ..., ik',..., iR-1]` where `ik + ik' = dk - 1` for all dimensions `k` in `dimensions`."

File: docs/status.md (hunk 6 above)
Comment (subhankarshah) [reply docs/status.md:115]: #273 
Created an issue to add verification for ReverseOp

File: docs/spec_draft.md (hunk 4 above)
Comment (burmako) [reply docs/spec_draft.md]: Oof you're right I didn't consider that.

I think that the canonical form (i.e. what's printed by -opt tools) just prints everything on a single line. I agree that this is hard to follow. 

This is why personally I've been using [a notation like this](https://github.com/tensorflow/mlir-hlo/blob/34028950c5fac34a074bb9da63659acea210bbcb/tests/Dialect/mhlo/hlo-legalize-to-stablehlo.mlir#L902-L911), but that's not standardized. What do you think about that format?

File: docs/status.md (hunk 6 above)
Comment (burmako) [reply docs/status.md]: Nice catch re: the missing verifier!

**Review by burmako — APPROVED** [review]

File: docs/spec_draft.md (hunk 4 above)
Comment (ghpvnist) [reply docs/spec_draft.md]: No objections for the format, I was implicitly following that format as well for ops that are too long to write in one line.

I'll also create a PR fixing existing ops with attributes to follow this format.

**Review by ghpvnist — APPROVED** [review]
